names parameter of pd.read_csv restricted to list[str] even though the valid types are much broader

**Describe the bug**
Closely related to #605

The stubs for the `read_csv` function, in `pandas-stubs/io/parsers/readers.pyi`, type hint the 'names' param as `list[str]`, even though a much wider range of types is usable here. The [pandas docs](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) say:

> names: array-like, optional
List of column names to use. If the file contains a header row, then you should explicitly pass header=0 to override the column names. Duplicates in this list are not allowed.

And the [pandas code](https://github.com/pandas-dev/pandas/blob/0ba82890fb398a7afc2d6a1b83e31fa3b33fceae/pandas/io/parsers/readers.py#L616) suggests a `Sequence[Hashable]` is allowed.

**To Reproduce**
1. Provide a minimal runnable `pandas` example that is not properly checked by the stubs.

```python
import pandas as pd

pd.read_csv("test.csv", names=("first", "second", "third"))
pd.read_csv("test.csv", names=(7, 8, 9))
# MultiIndex - although unclear to me if this is part of the 'public api' and therefore something pandas-stubs wants to cover
pd.read_csv("test.csv", names=(("head", "first"), ("tail", "second"), ("tail", "third")))
```

contents of `test.csv` was just:

```csv
1,2,3
4,5,6
```

3. mypy
4. Show the error message received from that type checker while checking your example.

abridged, because there's lots of output about possible overload variants:

```console
test_read_csv.py:3: error: No overload variant of "read_csv" matches argument types "str", "Tuple[str, str, str]"  [call-overload]
...
test_read_csv.py:4: error: No overload variant of "read_csv" matches argument types "str", "Tuple[int, int, int]"  [call-overload]
...
test_read_csv.py:6: error: No overload variant of "read_csv" matches argument types "str", "Tuple[Tuple[str, str], Tuple[str, str], Tuple[str, str]]"  [call-overload]
```

<details><summary>Full error</summary>

```console
test_read_csv.py:3: error: No overload variant of "read_csv" matches argument types "str", "Tuple[str, str, str]"  [call-overload]
test_read_csv.py:3: note: Possible overload variants:
test_read_csv.py:3: note:     def read_csv(filepath_or_buffer: Union[Union[str, PathLike[str]], ReadCsvBuffer[bytes], ReadCsvBuffer[str]], *, sep: Optional[str] = ..., delimiter: Optional[str] = ..., header: Union[int, Sequence[int], Literal['infer'], None] = ..., names: Optional[List[str]] = ..., index_col: Union[int, str, Sequence[Union[str, int]], Literal[False], None] = ..., usecols: Union[List[str], Tuple[str, ...], Sequence[int], Series[Any], Index, ndarray[Any, dtype[Any]], Callable[[str], bool], None] = ..., dtype: Union[Union[Union[ExtensionDtype, Union[str, dtype[generic], Type[str], Type[complex], Type[bool], Type[object]]], Dict[Any, Union[ExtensionDtype, Union[str, dtype[generic], Type[str], Type[complex], Type[bool], Type[object]]]]], defaultdict[Any, Any], None] = ..., engine: Optional[Literal['c', 'python', 'pyarrow', 'python-fwf']] = ..., converters: Union[Mapping[Union[int, str], Callable[[str], Any]], Mapping[int, Callable[[str], Any]], Mapping[str, Callable[[str], Any]], None] = ..., true_values: List[str] = ..., false_values: List[str] = ..., skipinitialspace: bool = ..., skiprows: Union[int, Sequence[int], Callable[[int], bool]] = ..., skipfooter: int = ..., nrows: Optional[int] = ..., na_values: Union[Sequence[str], Mapping[str, Sequence[str]]] = ..., keep_default_na: bool = ..., na_filter: bool = ..., verbose: bool = ..., skip_blank_lines: bool = ..., parse_dates: Union[bool, List[int], List[str], Sequence[Sequence[int]], Mapping[str, Sequence[Union[int, str]]]] = ..., infer_datetime_format: bool = ..., keep_date_col: bool = ..., date_parser: Callable[..., Any] = ..., dayfirst: bool = ..., cache_dates: bool = ..., iterator: Literal[True], chunksize: Optional[int] = ..., compression: Union[None, Dict[str, Any], Literal['infer', 'gzip', 'bz2', 'zip', 'xz', 'zstd']] = ..., thousands: Optional[str] = ..., decimal: str = ..., lineterminator: Optional[str] = ..., quotechar: str = ..., quoting: Literal[0, 1, 2, 3] = ..., doublequote: bool = ..., escapechar: Optional[str] = ..., comment: Optional[str] = ..., encoding: Optional[str] = ..., encoding_errors: Optional[str] = ..., dialect: Union[str, Dialect] = ..., on_bad_lines: Union[Callable[[List[str]], Optional[List[str]]], Literal['error', 'warn', 'skip']] = ..., delim_whitespace: bool = ..., low_memory: bool = ..., memory_map: bool = ..., float_precision: Optional[Literal['high', 'legacy', 'round_trip']] = ..., storage_options: Optional[Optional[Dict[str, Any]]] = ...) -> TextFileReader
test_read_csv.py:3: note:     def read_csv(filepath_or_buffer: Union[Union[str, PathLike[str]], ReadCsvBuffer[bytes], ReadCsvBuffer[str]], *, sep: Optional[str] = ..., delimiter: Optional[str] = ..., header: Union[int, Sequence[int], Literal['infer'], None] = ..., names: Optional[List[str]] = ..., index_col: Union[int, str, Sequence[Union[str, int]], Literal[False], None] = ..., usecols: Union[List[str], Tuple[str, ...], Sequence[int], Series[Any], Index, ndarray[Any, dtype[Any]], Callable[[str], bool], None] = ..., dtype: Union[Union[Union[ExtensionDtype, Union[str, dtype[generic], Type[str], Type[complex], Type[bool], Type[object]]], Dict[Any, Union[ExtensionDtype, Union[str, dtype[generic], Type[str], Type[complex], Type[bool], Type[object]]]]], defaultdict[Any, Any], None] = ..., engine: Optional[Literal['c', 'python', 'pyarrow', 'python-fwf']] = ..., converters: Union[Mapping[Union[int, str], Callable[[str], Any]], Mapping[int, Callable[[str], Any]], Mapping[str, Callable[[str], Any]], None] = ..., true_values: List[str] = ..., false_values: List[str] = ..., skipinitialspace: bool = ..., skiprows: Union[int, Sequence[int], Callable[[int], bool]] = ..., skipfooter: int = ..., nrows: Optional[int] = ..., na_values: Union[Sequence[str], Mapping[str, Sequence[str]]] = ..., keep_default_na: bool = ..., na_filter: bool = ..., verbose: bool = ..., skip_blank_lines: bool = ..., parse_dates: Union[bool, List[int], List[str], Sequence[Sequence[int]], Mapping[str, Sequence[Union[int, str]]]] = ..., infer_datetime_format: bool = ..., keep_date_col: bool = ..., date_parser: Callable[..., Any] = ..., dayfirst: bool = ..., cache_dates: bool = ..., iterator: bool = ..., chunksize: int, compression: Union[None, Dict[str, Any], Literal['infer', 'gzip', 'bz2', 'zip', 'xz', 'zstd']] = ..., thousands: Optional[str] = ..., decimal: str = ..., lineterminator: Optional[str] = ..., quotechar: str = ..., quoting: Literal[0, 1, 2, 3] = ..., doublequote: bool = ..., escapechar: Optional[str] = ..., comment: Optional[str] = ..., encoding: Optional[str] = ..., encoding_errors: Optional[str] = ..., dialect: Union[str, Dialect] = ..., on_bad_lines: Union[Callable[[List[str]], Optional[List[str]]], Literal['error', 'warn', 'skip']] = ..., delim_whitespace: bool = ..., low_memory: bool = ..., memory_map: bool = ..., float_precision: Optional[Literal['high', 'legacy', 'round_trip']] = ..., storage_options: Optional[Optional[Dict[str, Any]]] = ...) -> TextFileReader
test_read_csv.py:3: note:     def read_csv(filepath_or_buffer: Union[Union[str, PathLike[str]], ReadCsvBuffer[bytes], ReadCsvBuffer[str]], *, sep: Optional[str] = ..., delimiter: Optional[str] = ..., header: Union[int, Sequence[int], Literal['infer'], None] = ..., names: Optional[List[str]] = ..., index_col: Union[int, str, Sequence[Union[str, int]], Literal[False], None] = ..., usecols: Union[List[str], Tuple[str, ...], Sequence[int], Series[Any], Index, ndarray[Any, dtype[Any]], Callable[[str], bool], None] = ..., dtype: Union[Union[Union[ExtensionDtype, Union[str, dtype[generic], Type[str], Type[complex], Type[bool], Type[object]]], Dict[Any, Union[ExtensionDtype, Union[str, dtype[generic], Type[str], Type[complex], Type[bool], Type[object]]]]], defaultdict[Any, Any], None] = ..., engine: Optional[Literal['c', 'python', 'pyarrow', 'python-fwf']] = ..., converters: Union[Mapping[Union[int, str], Callable[[str], Any]], Mapping[int, Callable[[str], Any]], Mapping[str, Callable[[str], Any]], None] = ..., true_values: List[str] = ..., false_values: List[str] = ..., skipinitialspace: bool = ..., skiprows: Union[int, Sequence[int], Callable[[int], bool]] = ..., skipfooter: int = ..., nrows: Optional[int] = ..., na_values: Union[Sequence[str], Mapping[str, Sequence[str]]] = ..., keep_default_na: bool = ..., na_filter: bool = ..., verbose: bool = ..., skip_blank_lines: bool = ..., parse_dates: Union[bool, List[int], List[str], Sequence[Sequence[int]], Mapping[str, Sequence[Union[int, str]]]] = ..., infer_datetime_format: bool = ..., keep_date_col: bool = ..., date_parser: Callable[..., Any] = ..., dayfirst: bool = ..., cache_dates: bool = ..., iterator: Literal[False] = ..., chunksize: None = ..., compression: Union[None, Dict[str, Any], Literal['infer', 'gzip', 'bz2', 'zip', 'xz', 'zstd']] = ..., thousands: Optional[str] = ..., decimal: str = ..., lineterminator: Optional[str] = ..., quotechar: str = ..., quoting: Literal[0, 1, 2, 3] = ..., doublequote: bool = ..., escapechar: Optional[str] = ..., comment: Optional[str] = ..., encoding: Optional[str] = ..., encoding_errors: Optional[str] = ..., dialect: Union[str, Dialect] = ..., on_bad_lines: Union[Callable[[List[str]], Optional[List[str]]], Literal['error', 'warn', 'skip']] = ..., delim_whitespace: bool = ..., low_memory: bool = ..., memory_map: bool = ..., float_precision: Optional[Literal['high', 'legacy', 'round_trip']] = ..., storage_options: Optional[Optional[Dict[str, Any]]] = ...) -> DataFrame
test_read_csv.py:4: error: No overload variant of "read_csv" matches argument types "str", "Tuple[int, int, int]"  [call-overload]
test_read_csv.py:4: note: Possible overload variants:
test_read_csv.py:4: note:     def read_csv(filepath_or_buffer: Union[Union[str, PathLike[str]], ReadCsvBuffer[bytes], ReadCsvBuffer[str]], *, sep: Optional[str] = ..., delimiter: Optional[str] = ..., header: Union[int, Sequence[int], Literal['infer'], None] = ..., names: Optional[List[str]] = ..., index_col: Union[int, str, Sequence[Union[str, int]], Literal[False], None] = ..., usecols: Union[List[str], Tuple[str, ...], Sequence[int], Series[Any], Index, ndarray[Any, dtype[Any]], Callable[[str], bool], None] = ..., dtype: Union[Union[Union[ExtensionDtype, Union[str, dtype[generic], Type[str], Type[complex], Type[bool], Type[object]]], Dict[Any, Union[ExtensionDtype, Union[str, dtype[generic], Type[str], Type[complex], Type[bool], Type[object]]]]], defaultdict[Any, Any], None] = ..., engine: Optional[Literal['c', 'python', 'pyarrow', 'python-fwf']] = ..., converters: Union[Mapping[Union[int, str], Callable[[str], Any]], Mapping[int, Callable[[str], Any]], Mapping[str, Callable[[str], Any]], None] = ..., true_values: List[str] = ..., false_values: List[str] = ..., skipinitialspace: bool = ..., skiprows: Union[int, Sequence[int], Callable[[int], bool]] = ..., skipfooter: int = ..., nrows: Optional[int] = ..., na_values: Union[Sequence[str], Mapping[str, Sequence[str]]] = ..., keep_default_na: bool = ..., na_filter: bool = ..., verbose: bool = ..., skip_blank_lines: bool = ..., parse_dates: Union[bool, List[int], List[str], Sequence[Sequence[int]], Mapping[str, Sequence[Union[int, str]]]] = ..., infer_datetime_format: bool = ..., keep_date_col: bool = ..., date_parser: Callable[..., Any] = ..., dayfirst: bool = ..., cache_dates: bool = ..., iterator: Literal[True], chunksize: Optional[int] = ..., compression: Union[None, Dict[str, Any], Literal['infer', 'gzip', 'bz2', 'zip', 'xz', 'zstd']] = ..., thousands: Optional[str] = ..., decimal: str = ..., lineterminator: Optional[str] = ..., quotechar: str = ..., quoting: Literal[0, 1, 2, 3] = ..., doublequote: bool = ..., escapechar: Optional[str] = ..., comment: Optional[str] = ..., encoding: Optional[str] = ..., encoding_errors: Optional[str] = ..., dialect: Union[str, Dialect] = ..., on_bad_lines: Union[Callable[[List[str]], Optional[List[str]]], Literal['error', 'warn', 'skip']] = ..., delim_whitespace: bool = ..., low_memory: bool = ..., memory_map: bool = ..., float_precision: Optional[Literal['high', 'legacy', 'round_trip']] = ..., storage_options: Optional[Optional[Dict[str, Any]]] = ...) -> TextFileReader
test_read_csv.py:4: note:     def read_csv(filepath_or_buffer: Union[Union[str, PathLike[str]], ReadCsvBuffer[bytes], ReadCsvBuffer[str]], *, sep: Optional[str] = ..., delimiter: Optional[str] = ..., header: Union[int, Sequence[int], Literal['infer'], None] = ..., names: Optional[List[str]] = ..., index_col: Union[int, str, Sequence[Union[str, int]], Literal[False], None] = ..., usecols: Union[List[str], Tuple[str, ...], Sequence[int], Series[Any], Index, ndarray[Any, dtype[Any]], Callable[[str], bool], None] = ..., dtype: Union[Union[Union[ExtensionDtype, Union[str, dtype[generic], Type[str], Type[complex], Type[bool], Type[object]]], Dict[Any, Union[ExtensionDtype, Union[str, dtype[generic], Type[str], Type[complex], Type[bool], Type[object]]]]], defaultdict[Any, Any], None] = ..., engine: Optional[Literal['c', 'python', 'pyarrow', 'python-fwf']] = ..., converters: Union[Mapping[Union[int, str], Callable[[str], Any]], Mapping[int, Callable[[str], Any]], Mapping[str, Callable[[str], Any]], None] = ..., true_values: List[str] = ..., false_values: List[str] = ..., skipinitialspace: bool = ..., skiprows: Union[int, Sequence[int], Callable[[int], bool]] = ..., skipfooter: int = ..., nrows: Optional[int] = ..., na_values: Union[Sequence[str], Mapping[str, Sequence[str]]] = ..., keep_default_na: bool = ..., na_filter: bool = ..., verbose: bool = ..., skip_blank_lines: bool = ..., parse_dates: Union[bool, List[int], List[str], Sequence[Sequence[int]], Mapping[str, Sequence[Union[int, str]]]] = ..., infer_datetime_format: bool = ..., keep_date_col: bool = ..., date_parser: Callable[..., Any] = ..., dayfirst: bool = ..., cache_dates: bool = ..., iterator: bool = ..., chunksize: int, compression: Union[None, Dict[str, Any], Literal['infer', 'gzip', 'bz2', 'zip', 'xz', 'zstd']] = ..., thousands: Optional[str] = ..., decimal: str = ..., lineterminator: Optional[str] = ..., quotechar: str = ..., quoting: Literal[0, 1, 2, 3] = ..., doublequote: bool = ..., escapechar: Optional[str] = ..., comment: Optional[str] = ..., encoding: Optional[str] = ..., encoding_errors: Optional[str] = ..., dialect: Union[str, Dialect] = ..., on_bad_lines: Union[Callable[[List[str]], Optional[List[str]]], Literal['error', 'warn', 'skip']] = ..., delim_whitespace: bool = ..., low_memory: bool = ..., memory_map: bool = ..., float_precision: Optional[Literal['high', 'legacy', 'round_trip']] = ..., storage_options: Optional[Optional[Dict[str, Any]]] = ...) -> TextFileReader
test_read_csv.py:4: note:     def read_csv(filepath_or_buffer: Union[Union[str, PathLike[str]], ReadCsvBuffer[bytes], ReadCsvBuffer[str]], *, sep: Optional[str] = ..., delimiter: Optional[str] = ..., header: Union[int, Sequence[int], Literal['infer'], None] = ..., names: Optional[List[str]] = ..., index_col: Union[int, str, Sequence[Union[str, int]], Literal[False], None] = ..., usecols: Union[List[str], Tuple[str, ...], Sequence[int], Series[Any], Index, ndarray[Any, dtype[Any]], Callable[[str], bool], None] = ..., dtype: Union[Union[Union[ExtensionDtype, Union[str, dtype[generic], Type[str], Type[complex], Type[bool], Type[object]]], Dict[Any, Union[ExtensionDtype, Union[str, dtype[generic], Type[str], Type[complex], Type[bool], Type[object]]]]], defaultdict[Any, Any], None] = ..., engine: Optional[Literal['c', 'python', 'pyarrow', 'python-fwf']] = ..., converters: Union[Mapping[Union[int, str], Callable[[str], Any]], Mapping[int, Callable[[str], Any]], Mapping[str, Callable[[str], Any]], None] = ..., true_values: List[str] = ..., false_values: List[str] = ..., skipinitialspace: bool = ..., skiprows: Union[int, Sequence[int], Callable[[int], bool]] = ..., skipfooter: int = ..., nrows: Optional[int] = ..., na_values: Union[Sequence[str], Mapping[str, Sequence[str]]] = ..., keep_default_na: bool = ..., na_filter: bool = ..., verbose: bool = ..., skip_blank_lines: bool = ..., parse_dates: Union[bool, List[int], List[str], Sequence[Sequence[int]], Mapping[str, Sequence[Union[int, str]]]] = ..., infer_datetime_format: bool = ..., keep_date_col: bool = ..., date_parser: Callable[..., Any] = ..., dayfirst: bool = ..., cache_dates: bool = ..., iterator: Literal[False] = ..., chunksize: None = ..., compression: Union[None, Dict[str, Any], Literal['infer', 'gzip', 'bz2', 'zip', 'xz', 'zstd']] = ..., thousands: Optional[str] = ..., decimal: str = ..., lineterminator: Optional[str] = ..., quotechar: str = ..., quoting: Literal[0, 1, 2, 3] = ..., doublequote: bool = ..., escapechar: Optional[str] = ..., comment: Optional[str] = ..., encoding: Optional[str] = ..., encoding_errors: Optional[str] = ..., dialect: Union[str, Dialect] = ..., on_bad_lines: Union[Callable[[List[str]], Optional[List[str]]], Literal['error', 'warn', 'skip']] = ..., delim_whitespace: bool = ..., low_memory: bool = ..., memory_map: bool = ..., float_precision: Optional[Literal['high', 'legacy', 'round_trip']] = ..., storage_options: Optional[Optional[Dict[str, Any]]] = ...) -> DataFrame
test_read_csv.py:6: error: No overload variant of "read_csv" matches argument types "str", "Tuple[Tuple[str, str], Tuple[str, str], Tuple[str, str]]"  [call-overload]
test_read_csv.py:6: note: Possible overload variants:
test_read_csv.py:6: note:     def read_csv(filepath_or_buffer: Union[Union[str, PathLike[str]], ReadCsvBuffer[bytes], ReadCsvBuffer[str]], *, sep: Optional[str] = ..., delimiter: Optional[str] = ..., header: Union[int, Sequence[int], Literal['infer'], None] = ..., names: Optional[List[str]] = ..., index_col: Union[int, str, Sequence[Union[str, int]], Literal[False], None] = ..., usecols: Union[List[str], Tuple[str, ...], Sequence[int], Series[Any], Index, ndarray[Any, dtype[Any]], Callable[[str], bool], None] = ..., dtype: Union[Union[Union[ExtensionDtype, Union[str, dtype[generic], Type[str], Type[complex], Type[bool], Type[object]]], Dict[Any, Union[ExtensionDtype, Union[str, dtype[generic], Type[str], Type[complex], Type[bool], Type[object]]]]], defaultdict[Any, Any], None] = ..., engine: Optional[Literal['c', 'python', 'pyarrow', 'python-fwf']] = ..., converters: Union[Mapping[Union[int, str], Callable[[str], Any]], Mapping[int, Callable[[str], Any]], Mapping[str, Callable[[str], Any]], None] = ..., true_values: List[str] = ..., false_values: List[str] = ..., skipinitialspace: bool = ..., skiprows: Union[int, Sequence[int], Callable[[int], bool]] = ..., skipfooter: int = ..., nrows: Optional[int] = ..., na_values: Union[Sequence[str], Mapping[str, Sequence[str]]] = ..., keep_default_na: bool = ..., na_filter: bool = ..., verbose: bool = ..., skip_blank_lines: bool = ..., parse_dates: Union[bool, List[int], List[str], Sequence[Sequence[int]], Mapping[str, Sequence[Union[int, str]]]] = ..., infer_datetime_format: bool = ..., keep_date_col: bool = ..., date_parser: Callable[..., Any] = ..., dayfirst: bool = ..., cache_dates: bool = ..., iterator: Literal[True], chunksize: Optional[int] = ..., compression: Union[None, Dict[str, Any], Literal['infer', 'gzip', 'bz2', 'zip', 'xz', 'zstd']] = ..., thousands: Optional[str] = ..., decimal: str = ..., lineterminator: Optional[str] = ..., quotechar: str = ..., quoting: Literal[0, 1, 2, 3] = ..., doublequote: bool = ..., escapechar: Optional[str] = ..., comment: Optional[str] = ..., encoding: Optional[str] = ..., encoding_errors: Optional[str] = ..., dialect: Union[str, Dialect] = ..., on_bad_lines: Union[Callable[[List[str]], Optional[List[str]]], Literal['error', 'warn', 'skip']] = ..., delim_whitespace: bool = ..., low_memory: bool = ..., memory_map: bool = ..., float_precision: Optional[Literal['high', 'legacy', 'round_trip']] = ..., storage_options: Optional[Optional[Dict[str, Any]]] = ...) -> TextFileReader
test_read_csv.py:6: note:     def read_csv(filepath_or_buffer: Union[Union[str, PathLike[str]], ReadCsvBuffer[bytes], ReadCsvBuffer[str]], *, sep: Optional[str] = ..., delimiter: Optional[str] = ..., header: Union[int, Sequence[int], Literal['infer'], None] = ..., names: Optional[List[str]] = ..., index_col: Union[int, str, Sequence[Union[str, int]], Literal[False], None] = ..., usecols: Union[List[str], Tuple[str, ...], Sequence[int], Series[Any], Index, ndarray[Any, dtype[Any]], Callable[[str], bool], None] = ..., dtype: Union[Union[Union[ExtensionDtype, Union[str, dtype[generic], Type[str], Type[complex], Type[bool], Type[object]]], Dict[Any, Union[ExtensionDtype, Union[str, dtype[generic], Type[str], Type[complex], Type[bool], Type[object]]]]], defaultdict[Any, Any], None] = ..., engine: Optional[Literal['c', 'python', 'pyarrow', 'python-fwf']] = ..., converters: Union[Mapping[Union[int, str], Callable[[str], Any]], Mapping[int, Callable[[str], Any]], Mapping[str, Callable[[str], Any]], None] = ..., true_values: List[str] = ..., false_values: List[str] = ..., skipinitialspace: bool = ..., skiprows: Union[int, Sequence[int], Callable[[int], bool]] = ..., skipfooter: int = ..., nrows: Optional[int] = ..., na_values: Union[Sequence[str], Mapping[str, Sequence[str]]] = ..., keep_default_na: bool = ..., na_filter: bool = ..., verbose: bool = ..., skip_blank_lines: bool = ..., parse_dates: Union[bool, List[int], List[str], Sequence[Sequence[int]], Mapping[str, Sequence[Union[int, str]]]] = ..., infer_datetime_format: bool = ..., keep_date_col: bool = ..., date_parser: Callable[..., Any] = ..., dayfirst: bool = ..., cache_dates: bool = ..., iterator: bool = ..., chunksize: int, compression: Union[None, Dict[str, Any], Literal['infer', 'gzip', 'bz2', 'zip', 'xz', 'zstd']] = ..., thousands: Optional[str] = ..., decimal: str = ..., lineterminator: Optional[str] = ..., quotechar: str = ..., quoting: Literal[0, 1, 2, 3] = ..., doublequote: bool = ..., escapechar: Optional[str] = ..., comment: Optional[str] = ..., encoding: Optional[str] = ..., encoding_errors: Optional[str] = ..., dialect: Union[str, Dialect] = ..., on_bad_lines: Union[Callable[[List[str]], Optional[List[str]]], Literal['error', 'warn', 'skip']] = ..., delim_whitespace: bool = ..., low_memory: bool = ..., memory_map: bool = ..., float_precision: Optional[Literal['high', 'legacy', 'round_trip']] = ..., storage_options: Optional[Optional[Dict[str, Any]]] = ...) -> TextFileReader
test_read_csv.py:6: note:     def read_csv(filepath_or_buffer: Union[Union[str, PathLike[str]], ReadCsvBuffer[bytes], ReadCsvBuffer[str]], *, sep: Optional[str] = ..., delimiter: Optional[str] = ..., header: Union[int, Sequence[int], Literal['infer'], None] = ..., names: Optional[List[str]] = ..., index_col: Union[int, str, Sequence[Union[str, int]], Literal[False], None] = ..., usecols: Union[List[str], Tuple[str, ...], Sequence[int], Series[Any], Index, ndarray[Any, dtype[Any]], Callable[[str], bool], None] = ..., dtype: Union[Union[Union[ExtensionDtype, Union[str, dtype[generic], Type[str], Type[complex], Type[bool], Type[object]]], Dict[Any, Union[ExtensionDtype, Union[str, dtype[generic], Type[str], Type[complex], Type[bool], Type[object]]]]], defaultdict[Any, Any], None] = ..., engine: Optional[Literal['c', 'python', 'pyarrow', 'python-fwf']] = ..., converters: Union[Mapping[Union[int, str], Callable[[str], Any]], Mapping[int, Callable[[str], Any]], Mapping[str, Callable[[str], Any]], None] = ..., true_values: List[str] = ..., false_values: List[str] = ..., skipinitialspace: bool = ..., skiprows: Union[int, Sequence[int], Callable[[int], bool]] = ..., skipfooter: int = ..., nrows: Optional[int] = ..., na_values: Union[Sequence[str], Mapping[str, Sequence[str]]] = ..., keep_default_na: bool = ..., na_filter: bool = ..., verbose: bool = ..., skip_blank_lines: bool = ..., parse_dates: Union[bool, List[int], List[str], Sequence[Sequence[int]], Mapping[str, Sequence[Union[int, str]]]] = ..., infer_datetime_format: bool = ..., keep_date_col: bool = ..., date_parser: Callable[..., Any] = ..., dayfirst: bool = ..., cache_dates: bool = ..., iterator: Literal[False] = ..., chunksize: None = ..., compression: Union[None, Dict[str, Any], Literal['infer', 'gzip', 'bz2', 'zip', 'xz', 'zstd']] = ..., thousands: Optional[str] = ..., decimal: str = ..., lineterminator: Optional[str] = ..., quotechar: str = ..., quoting: Literal[0, 1, 2, 3] = ..., doublequote: bool = ..., escapechar: Optional[str] = ..., comment: Optional[str] = ..., encoding: Optional[str] = ..., encoding_errors: Optional[str] = ..., dialect: Union[str, Dialect] = ..., on_bad_lines: Union[Callable[[List[str]], Optional[List[str]]], Literal['error', 'warn', 'skip']] = ..., delim_whitespace: bool = ..., low_memory: bool = ..., memory_map: bool = ..., float_precision: Optional[Literal['high', 'legacy', 'round_trip']] = ..., storage_options: Optional[Optional[Dict[str, Any]]] = ...) -> DataFrame
Found 3 errors in 1 file (checked 1 source file)
```

</details> 

**Please complete the following information:**
 - OS: MacOS
 - OS Version: Ventura 13.3 (22E252)
 - python version: 3.10.0
 - version of type checker: mypy 1.1.1 (compiled: yes)
 - version of installed `pandas-stubs`


**Additional context**
I was planning to open a PR for this, but I'm actually unsure what type hint is best. My thought process went:

1. Maybe it can be a `Sequence[Hashable]` like in pandas-dev/pandas.
2. No, it can't because this wouldn't disallow a single string (which errors - `pd.read_csv("test.csv", names="abc")` gives `ValueError: Names should be an ordered collection.`).
3. I can take inspiration from `ListLikeExceptSeriesAndStr` - but a `Series` is valid too, and it needs to be Hashables only inside the ListLikes... what about `ListLikeExceptStrHashable = TypeVar("ListLikeExceptStrHashable", MutableSequence[Hashable], np.ndarray, tuple[Hashable], "Index")`
4. That doesn't work because `pd.Index(({1: 1, 2:2, 3:3}, {"test": "this"}))` is possible to construct, but errors when trying to use it as the `names` param for `read_csv`, because it's not hashable (gives a `TypeError: unhashable type: 'dict'`).

I could just exclude the `Index` option, but that seems like a shame. However, I don't think there's an easy way of type hinting 'it's an Index, and the elements in the Index are Hashable', but maybe I'm missing something.

If I had advice here on an acceptable type I'd happily provide a PR - I think allowing: MutableSequence[Hashable], np.ndarray, tuple[Hashable] would be an improvement over what we have now even if it's not perfect.

--- a/pandas-stubs/_typing.pyi
+++ b/pandas-stubs/_typing.pyi
@@ -298,6 +298,16 @@ np_ndarray_str: TypeAlias = npt.NDArray[np.str_]
 
 IndexType: TypeAlias = slice | np_ndarray_anyint | Index | list[int] | Series[int]
 MaskType: TypeAlias = Series[bool] | np_ndarray_bool | list[bool]
+UsecolsArgType: TypeAlias = (
+    MutableSequence[str]
+    | tuple[str, ...]
+    | Sequence[int]
+    | Series
+    | Index
+    | np.ndarray
+    | Callable[[HashableT], bool]
+    | None
+)
 # Scratch types for generics
 S1 = TypeVar(
     "S1",

--- a/pandas-stubs/_typing.pyi
+++ b/pandas-stubs/_typing.pyi
@@ -269,6 +269,9 @@ ListLikeExceptSeriesAndStr = TypeVar(
     "ListLikeExceptSeriesAndStr", MutableSequence, np.ndarray, tuple, "Index"
 )
 ListLikeU: TypeAlias = Sequence | np.ndarray | Series | Index
+ListLikeHashable: TypeAlias = (
+    MutableSequence[HashableT] | np.ndarray | tuple[HashableT, ...] | range
+)
 StrLike: TypeAlias = str | np.str_
 IndexIterScalar: TypeAlias = (
     str

--- a/pandas-stubs/io/clipboards.pyi
+++ b/pandas-stubs/io/clipboards.pyi
@@ -19,6 +19,7 @@ from pandas._typing import (
     CSVEngine,
     CSVQuoting,
     DtypeArg,
+    ListLikeHashable,
     StorageOptions,
     npt,
 )
@@ -31,7 +32,7 @@ def read_clipboard(
     *,
     delimiter: str | None = ...,
     header: int | Sequence[int] | Literal["infer"] | None = ...,
-    names: list[str] | None = ...,
+    names: ListLikeHashable | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
     usecols: list[str]
     | Sequence[int]
@@ -94,7 +95,7 @@ def read_clipboard(
     *,
     delimiter: str | None = ...,
     header: int | Sequence[int] | Literal["infer"] | None = ...,
-    names: list[str] | None = ...,
+    names: ListLikeHashable | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
     usecols: list[str]
     | Sequence[int]
@@ -157,7 +158,7 @@ def read_clipboard(
     *,
     delimiter: str | None = ...,
     header: int | Sequence[int] | Literal["infer"] | None = ...,
-    names: list[str] | None = ...,
+    names: ListLikeHashable | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
     usecols: list[str]
     | Sequence[int]

--- a/pandas-stubs/io/clipboards.pyi
+++ b/pandas-stubs/io/clipboards.pyi
@@ -11,8 +11,6 @@ from typing import (
 )
 
 from pandas.core.frame import DataFrame
-from pandas.core.indexes.base import Index
-from pandas.core.series import Series
 
 from pandas._typing import (
     CompressionOptions,
@@ -21,7 +19,7 @@ from pandas._typing import (
     DtypeArg,
     ListLikeHashable,
     StorageOptions,
-    npt,
+    UsecolsArgType,
 )
 
 from pandas.io.parsers import TextFileReader
@@ -34,13 +32,7 @@ def read_clipboard(
     header: int | Sequence[int] | Literal["infer"] | None = ...,
     names: ListLikeHashable | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
-    usecols: list[str]
-    | Sequence[int]
-    | Series
-    | Index
-    | npt.NDArray
-    | Callable[[str], bool]
-    | None = ...,
+    usecols: UsecolsArgType = ...,
     dtype: DtypeArg | defaultdict | None = ...,
     engine: CSVEngine | None = ...,
     converters: dict[int | str, Callable[[str], Any]] = ...,
@@ -97,13 +89,7 @@ def read_clipboard(
     header: int | Sequence[int] | Literal["infer"] | None = ...,
     names: ListLikeHashable | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
-    usecols: list[str]
-    | Sequence[int]
-    | Series
-    | Index
-    | npt.NDArray
-    | Callable[[str], bool]
-    | None = ...,
+    usecols: UsecolsArgType = ...,
     dtype: DtypeArg | defaultdict | None = ...,
     engine: CSVEngine | None = ...,
     converters: dict[int | str, Callable[[str], Any]] = ...,
@@ -160,13 +146,7 @@ def read_clipboard(
     header: int | Sequence[int] | Literal["infer"] | None = ...,
     names: ListLikeHashable | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
-    usecols: list[str]
-    | Sequence[int]
-    | Series
-    | Index
-    | npt.NDArray
-    | Callable[[str], bool]
-    | None = ...,
+    usecols: UsecolsArgType = ...,
     dtype: DtypeArg | defaultdict | None = ...,
     engine: CSVEngine | None = ...,
     converters: dict[int | str, Callable[[str], Any]] = ...,

--- a/pandas-stubs/io/clipboards.pyi
+++ b/pandas-stubs/io/clipboards.pyi
@@ -31,7 +31,7 @@ def read_clipboard(
     *,
     delimiter: str | None = ...,
     header: int | Sequence[int] | Literal["infer"] | None = ...,
-    names: list[str] = ...,
+    names: list[str] | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
     usecols: list[str]
     | Sequence[int]
@@ -94,7 +94,7 @@ def read_clipboard(
     *,
     delimiter: str | None = ...,
     header: int | Sequence[int] | Literal["infer"] | None = ...,
-    names: list[str] = ...,
+    names: list[str] | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
     usecols: list[str]
     | Sequence[int]
@@ -157,7 +157,7 @@ def read_clipboard(
     *,
     delimiter: str | None = ...,
     header: int | Sequence[int] | Literal["infer"] | None = ...,
-    names: list[str] = ...,
+    names: list[str] | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
     usecols: list[str]
     | Sequence[int]

--- a/pandas-stubs/io/excel/_base.pyi
+++ b/pandas-stubs/io/excel/_base.pyi
@@ -22,6 +22,7 @@ from xlrd.book import Book
 from pandas._typing import (
     Dtype,
     FilePath,
+    ListLikeHashable,
     ReadBuffer,
     StorageOptions,
     WriteExcelBuffer,
@@ -40,7 +41,7 @@ def read_excel(
     sheet_name: list[int | str] | None,
     *,
     header: int | Sequence[int] | None = ...,
-    names: list[str] | None = ...,
+    names: ListLikeHashable | None = ...,
     index_col: int | Sequence[int] | None = ...,
     usecols: Sequence[int] | Sequence[str] | Callable[[str], bool] | None = ...,
     dtype: str | Dtype | Mapping[str, str | Dtype] | None = ...,
@@ -78,7 +79,7 @@ def read_excel(
     sheet_name: int | str = ...,
     *,
     header: int | Sequence[int] | None = ...,
-    names: list[str] | None = ...,
+    names: ListLikeHashable | None = ...,
     index_col: int | Sequence[int] | None = ...,
     usecols: Sequence[int] | Sequence[str] | Callable[[str], bool] | None = ...,
     dtype: str | Dtype | Mapping[str, str | Dtype] | None = ...,
@@ -155,7 +156,7 @@ class ExcelFile:
         self,
         sheet_name: list[int | str] | None,
         header: int | Sequence[int] | None = ...,
-        names: list[str] | None = ...,
+        names: ListLikeHashable | None = ...,
         index_col: int | Sequence[int] | None = ...,
         usecols: str
         | Sequence[int]
@@ -185,7 +186,7 @@ class ExcelFile:
         self,
         sheet_name: int | str,
         header: int | Sequence[int] | None = ...,
-        names: list[str] | None = ...,
+        names: ListLikeHashable | None = ...,
         index_col: int | Sequence[int] | None = ...,
         usecols: str
         | Sequence[int]

--- a/pandas-stubs/io/excel/_base.pyi
+++ b/pandas-stubs/io/excel/_base.pyi
@@ -25,6 +25,7 @@ from pandas._typing import (
     ListLikeHashable,
     ReadBuffer,
     StorageOptions,
+    UsecolsArgType,
     WriteExcelBuffer,
 )
 
@@ -43,7 +44,7 @@ def read_excel(
     header: int | Sequence[int] | None = ...,
     names: ListLikeHashable | None = ...,
     index_col: int | Sequence[int] | None = ...,
-    usecols: Sequence[int] | Sequence[str] | Callable[[str], bool] | None = ...,
+    usecols: str | UsecolsArgType = ...,
     dtype: str | Dtype | Mapping[str, str | Dtype] | None = ...,
     engine: Literal["xlrd", "openpyxl", "odf", "pyxlsb"] | None = ...,
     converters: Mapping[int | str, Callable[[object], object]] | None = ...,
@@ -81,7 +82,7 @@ def read_excel(
     header: int | Sequence[int] | None = ...,
     names: ListLikeHashable | None = ...,
     index_col: int | Sequence[int] | None = ...,
-    usecols: Sequence[int] | Sequence[str] | Callable[[str], bool] | None = ...,
+    usecols: str | UsecolsArgType = ...,
     dtype: str | Dtype | Mapping[str, str | Dtype] | None = ...,
     engine: Literal["xlrd", "openpyxl", "odf", "pyxlsb"] | None = ...,
     converters: Mapping[int | str, Callable[[object], object]] | None = ...,
@@ -158,11 +159,7 @@ class ExcelFile:
         header: int | Sequence[int] | None = ...,
         names: ListLikeHashable | None = ...,
         index_col: int | Sequence[int] | None = ...,
-        usecols: str
-        | Sequence[int]
-        | Sequence[str]
-        | Callable[[str], bool]
-        | None = ...,
+        usecols: str | UsecolsArgType = ...,
         converters: dict[int | str, Callable[[object], object]] | None = ...,
         true_values: Iterable[Hashable] | None = ...,
         false_values: Iterable[Hashable] | None = ...,
@@ -188,11 +185,7 @@ class ExcelFile:
         header: int | Sequence[int] | None = ...,
         names: ListLikeHashable | None = ...,
         index_col: int | Sequence[int] | None = ...,
-        usecols: str
-        | Sequence[int]
-        | Sequence[str]
-        | Callable[[str], bool]
-        | None = ...,
+        usecols: str | UsecolsArgType = ...,
         converters: dict[int | str, Callable[[object], object]] | None = ...,
         true_values: Iterable[Hashable] | None = ...,
         false_values: Iterable[Hashable] | None = ...,

--- a/pandas-stubs/io/parsers/readers.pyi
+++ b/pandas-stubs/io/parsers/readers.pyi
@@ -16,8 +16,6 @@ from typing import (
 )
 
 from pandas.core.frame import DataFrame
-from pandas.core.indexes.base import Index
-from pandas.core.series import Series
 from typing_extensions import Self
 
 from pandas._typing import (
@@ -29,7 +27,7 @@ from pandas._typing import (
     ListLikeHashable,
     ReadCsvBuffer,
     StorageOptions,
-    npt,
+    UsecolsArgType,
 )
 
 from pandas.io.common import IOHandles
@@ -43,14 +41,7 @@ def read_csv(
     header: int | Sequence[int] | Literal["infer"] | None = ...,
     names: ListLikeHashable | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
-    usecols: list[str]
-    | tuple[str, ...]
-    | Sequence[int]
-    | Series
-    | Index
-    | npt.NDArray
-    | Callable[[str], bool]
-    | None = ...,
+    usecols: UsecolsArgType = ...,
     dtype: DtypeArg | defaultdict | None = ...,
     engine: CSVEngine | None = ...,
     converters: Mapping[int | str, Callable[[str], Any]]
@@ -109,14 +100,7 @@ def read_csv(
     header: int | Sequence[int] | Literal["infer"] | None = ...,
     names: ListLikeHashable | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
-    usecols: list[str]
-    | tuple[str, ...]
-    | Sequence[int]
-    | Series
-    | Index
-    | npt.NDArray
-    | Callable[[str], bool]
-    | None = ...,
+    usecols: UsecolsArgType = ...,
     dtype: DtypeArg | defaultdict | None = ...,
     engine: CSVEngine | None = ...,
     converters: Mapping[int | str, Callable[[str], Any]]
@@ -175,14 +159,7 @@ def read_csv(
     header: int | Sequence[int] | Literal["infer"] | None = ...,
     names: ListLikeHashable | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
-    usecols: list[str]
-    | tuple[str, ...]
-    | Sequence[int]
-    | Series
-    | Index
-    | npt.NDArray
-    | Callable[[str], bool]
-    | None = ...,
+    usecols: UsecolsArgType = ...,
     dtype: DtypeArg | defaultdict | None = ...,
     engine: CSVEngine | None = ...,
     converters: Mapping[int | str, Callable[[str], Any]]
@@ -241,14 +218,7 @@ def read_table(
     header: int | Sequence[int] | Literal["infer"] | None = ...,
     names: ListLikeHashable | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
-    usecols: list[str]
-    | tuple[str, ...]
-    | Sequence[int]
-    | Series
-    | Index
-    | npt.NDArray
-    | Callable[[str], bool]
-    | None = ...,
+    usecols: UsecolsArgType = ...,
     dtype: DtypeArg | defaultdict | None = ...,
     engine: CSVEngine | None = ...,
     converters: Mapping[int | str, Callable[[str], Any]]
@@ -307,14 +277,7 @@ def read_table(
     header: int | Sequence[int] | Literal["infer"] | None = ...,
     names: ListLikeHashable | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
-    usecols: list[str]
-    | tuple[str, ...]
-    | Sequence[int]
-    | Series
-    | Index
-    | npt.NDArray
-    | Callable[[str], bool]
-    | None = ...,
+    usecols: UsecolsArgType = ...,
     dtype: DtypeArg | defaultdict | None = ...,
     engine: CSVEngine | None = ...,
     converters: Mapping[int | str, Callable[[str], Any]]
@@ -373,14 +336,7 @@ def read_table(
     header: int | Sequence[int] | Literal["infer"] | None = ...,
     names: ListLikeHashable | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
-    usecols: list[str]
-    | tuple[str, ...]
-    | Sequence[int]
-    | Series
-    | Index
-    | npt.NDArray
-    | Callable[[str], bool]
-    | None = ...,
+    usecols: UsecolsArgType = ...,
     dtype: DtypeArg | defaultdict | None = ...,
     engine: CSVEngine | None = ...,
     converters: Mapping[int | str, Callable[[str], Any]]

--- a/pandas-stubs/io/parsers/readers.pyi
+++ b/pandas-stubs/io/parsers/readers.pyi
@@ -26,6 +26,7 @@ from pandas._typing import (
     CSVQuoting,
     DtypeArg,
     FilePath,
+    ListLikeHashable,
     ReadCsvBuffer,
     StorageOptions,
     npt,
@@ -40,7 +41,7 @@ def read_csv(
     sep: str | None = ...,
     delimiter: str | None = ...,
     header: int | Sequence[int] | Literal["infer"] | None = ...,
-    names: list[str] | None = ...,
+    names: ListLikeHashable | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
     usecols: list[str]
     | tuple[str, ...]
@@ -106,7 +107,7 @@ def read_csv(
     sep: str | None = ...,
     delimiter: str | None = ...,
     header: int | Sequence[int] | Literal["infer"] | None = ...,
-    names: list[str] | None = ...,
+    names: ListLikeHashable | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
     usecols: list[str]
     | tuple[str, ...]
@@ -172,7 +173,7 @@ def read_csv(
     sep: str | None = ...,
     delimiter: str | None = ...,
     header: int | Sequence[int] | Literal["infer"] | None = ...,
-    names: list[str] | None = ...,
+    names: ListLikeHashable | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
     usecols: list[str]
     | tuple[str, ...]
@@ -238,7 +239,7 @@ def read_table(
     sep: str | None = ...,
     delimiter: str | None = ...,
     header: int | Sequence[int] | Literal["infer"] | None = ...,
-    names: list[str] | None = ...,
+    names: ListLikeHashable | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
     usecols: list[str]
     | tuple[str, ...]
@@ -304,7 +305,7 @@ def read_table(
     sep: str | None = ...,
     delimiter: str | None = ...,
     header: int | Sequence[int] | Literal["infer"] | None = ...,
-    names: list[str] | None = ...,
+    names: ListLikeHashable | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
     usecols: list[str]
     | tuple[str, ...]
@@ -370,7 +371,7 @@ def read_table(
     sep: str | None = ...,
     delimiter: str | None = ...,
     header: int | Sequence[int] | Literal["infer"] | None = ...,
-    names: list[str] | None = ...,
+    names: ListLikeHashable | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
     usecols: list[str]
     | tuple[str, ...]

--- a/tests/test_io.py
+++ b/tests/test_io.py
@@ -12,6 +12,7 @@ from typing import (
     Generator,
     List,
     Literal,
+    Tuple,
     Union,
 )
 
@@ -206,6 +207,18 @@ def test_read_stata_iterator():
         check(assert_type(reader, StataReader), StataReader)
 
 
+def _true_if_b(s: str) -> bool:
+    return s == "b"
+
+
+def _true_if_greater_than_0(i: int) -> bool:
+    return i > 0
+
+
+def _true_if_first_param_is_head(t: Tuple[str, int]) -> bool:
+    return t[0] == "head"
+
+
 def test_clipboard():
     try:
         DF.to_clipboard()
@@ -228,6 +241,47 @@ def test_clipboard():
     check(
         assert_type(
             read_clipboard(names=(("first", 1), ("last", 2)), header=0), DataFrame
+        ),
+        DataFrame,
+    )
+    check(
+        assert_type(read_clipboard(usecols=None), DataFrame),
+        DataFrame,
+    )
+    check(
+        assert_type(read_clipboard(usecols=["a"]), DataFrame),
+        DataFrame,
+    )
+    check(
+        assert_type(read_clipboard(usecols=(0,)), DataFrame),
+        DataFrame,
+    )
+    check(
+        assert_type(read_clipboard(usecols=range(1)), DataFrame),
+        DataFrame,
+    )
+    check(
+        assert_type(read_clipboard(usecols=_true_if_b), DataFrame),
+        DataFrame,
+    )
+    check(
+        assert_type(
+            read_clipboard(
+                names=[1, 2], usecols=_true_if_greater_than_0, header=0, index_col=0
+            ),
+            DataFrame,
+        ),
+        DataFrame,
+    )
+    check(
+        assert_type(
+            read_clipboard(
+                names=(("head", 1), ("tail", 2)),
+                usecols=_true_if_first_param_is_head,
+                header=0,
+                index_col=0,
+            ),
+            DataFrame,
         ),
         DataFrame,
     )
@@ -501,6 +555,10 @@ def test_read_csv_iterator():
         tfr2.close()
 
 
+def _true_if_col1(s: str) -> bool:
+    return s == "col1"
+
+
 def test_types_read_csv() -> None:
     df = pd.DataFrame(data={"col1": [1, 2], "col2": [3, 4]})
     csv_df: str = df.to_csv()
@@ -543,6 +601,21 @@ def test_types_read_csv() -> None:
                 ("last", 2),
             ),
             header=0,
+        )
+        df19: pd.DataFrame = pd.read_csv(path, usecols=None)
+        df20: pd.DataFrame = pd.read_csv(path, usecols=["col1"])
+        df21: pd.DataFrame = pd.read_csv(path, usecols=(0,))
+        df22: pd.DataFrame = pd.read_csv(path, usecols=range(1))
+        df23: pd.DataFrame = pd.read_csv(path, usecols=_true_if_col1)
+        df24: pd.DataFrame = pd.read_csv(
+            path, names=[1, 2], usecols=_true_if_greater_than_0, header=0, index_col=0
+        )
+        df25: pd.DataFrame = pd.read_csv(
+            path,
+            names=(("head", 1), ("tail", 2)),
+            usecols=_true_if_first_param_is_head,
+            header=0,
+            index_col=0,
         )
 
         tfr1: TextFileReader = pd.read_csv(path, nrows=2, iterator=True, chunksize=3)
@@ -594,6 +667,67 @@ def test_read_table():
                         ("last", 2),
                     ),
                     header=0,
+                ),
+                DataFrame,
+            ),
+            DataFrame,
+        )
+        check(
+            assert_type(
+                read_table(path, usecols=None),
+                DataFrame,
+            ),
+            DataFrame,
+        )
+        check(
+            assert_type(
+                read_table(path, usecols=["a"]),
+                DataFrame,
+            ),
+            DataFrame,
+        )
+        check(
+            assert_type(
+                read_table(path, usecols=(0,)),
+                DataFrame,
+            ),
+            DataFrame,
+        )
+        check(
+            assert_type(
+                read_table(path, usecols=range(1)),
+                DataFrame,
+            ),
+            DataFrame,
+        )
+        check(
+            assert_type(
+                read_table(path, usecols=_true_if_b),
+                DataFrame,
+            ),
+            DataFrame,
+        )
+        check(
+            assert_type(
+                read_table(
+                    path,
+                    names=[1, 2],
+                    usecols=_true_if_greater_than_0,
+                    header=0,
+                    index_col=0,
+                ),
+                DataFrame,
+            ),
+            DataFrame,
+        )
+        check(
+            assert_type(
+                read_table(
+                    path,
+                    names=(("head", 1), ("tail", 2)),
+                    usecols=_true_if_first_param_is_head,
+                    header=0,
+                    index_col=0,
                 ),
                 DataFrame,
             ),
@@ -727,6 +861,52 @@ def test_read_excel() -> None:
         ),
         check(
             assert_type(pd.read_excel(path, names=range(1), header=0), pd.DataFrame),
+            pd.DataFrame,
+        )
+        check(
+            assert_type(pd.read_excel(path, usecols=None), pd.DataFrame),
+            pd.DataFrame,
+        )
+        check(
+            assert_type(pd.read_excel(path, usecols=["A"]), pd.DataFrame),
+            pd.DataFrame,
+        )
+        check(
+            assert_type(pd.read_excel(path, usecols=(0,)), pd.DataFrame),
+            pd.DataFrame,
+        )
+        check(
+            assert_type(pd.read_excel(path, usecols=range(1)), pd.DataFrame),
+            pd.DataFrame,
+        )
+        check(
+            assert_type(pd.read_excel(path, usecols=_true_if_b), pd.DataFrame),
+            pd.DataFrame,
+        )
+        check(
+            assert_type(
+                pd.read_excel(
+                    path,
+                    names=[1, 2],
+                    usecols=_true_if_greater_than_0,
+                    header=0,
+                    index_col=0,
+                ),
+                pd.DataFrame,
+            ),
+            pd.DataFrame,
+        )
+        check(
+            assert_type(
+                pd.read_excel(
+                    path,
+                    names=(("head", 1), ("tail", 2)),
+                    usecols=_true_if_first_param_is_head,
+                    header=0,
+                    index_col=0,
+                ),
+                pd.DataFrame,
+            ),
             pd.DataFrame,
         )
 

--- a/tests/test_io.py
+++ b/tests/test_io.py
@@ -517,6 +517,20 @@ def test_types_read_csv() -> None:
         df12: pd.DataFrame = pd.read_csv(path, usecols=("col1",))
         df13: pd.DataFrame = pd.read_csv(path, usecols=pd.Series(data=["col1"]))
         df14: pd.DataFrame = pd.read_csv(path, converters=None)
+        df15: pd.DataFrame = pd.read_csv(path, names=("first", "second"), header=0)
+        df16: pd.DataFrame = pd.read_csv(path, names=range(2), header=0)
+        df17: pd.DataFrame = pd.read_csv(path, names=(1, "two"), header=0)
+        df18: pd.DataFrame = pd.read_csv(
+            path,
+            names=(
+                (
+                    "first",
+                    1,
+                ),
+                ("last", 2),
+            ),
+            header=0,
+        )
 
         tfr1: TextFileReader = pd.read_csv(path, nrows=2, iterator=True, chunksize=3)
         tfr1.close()
@@ -539,6 +553,37 @@ def test_read_table():
         check(assert_type(read_table(path, chunksize=None), DataFrame), DataFrame)
         check(
             assert_type(read_table(path, dtype=defaultdict(lambda: "f8")), DataFrame),
+            DataFrame,
+        )
+        check(
+            assert_type(
+                read_table(path, names=("first", "second"), header=0), DataFrame
+            ),
+            DataFrame,
+        )
+        check(
+            assert_type(read_table(path, names=range(2), header=0), DataFrame),
+            DataFrame,
+        )
+        check(
+            assert_type(read_table(path, names=(1, "two"), header=0), DataFrame),
+            DataFrame,
+        )
+        check(
+            assert_type(
+                read_table(
+                    path,
+                    names=(
+                        (
+                            "first",
+                            1,
+                        ),
+                        ("last", 2),
+                    ),
+                    header=0,
+                ),
+                DataFrame,
+            ),
             DataFrame,
         )
 

--- a/tests/test_io.py
+++ b/tests/test_io.py
@@ -710,6 +710,25 @@ def test_read_excel() -> None:
             ),
             dict,
         )
+        check(
+            assert_type(pd.read_excel(path, names=("test",), header=0), pd.DataFrame),
+            pd.DataFrame,
+        )
+        check(
+            assert_type(pd.read_excel(path, names=(1,), header=0), pd.DataFrame),
+            pd.DataFrame,
+        )
+        check(
+            assert_type(
+                pd.read_excel(path, names=(("higher", "lower"),), header=0),
+                pd.DataFrame,
+            ),
+            pd.DataFrame,
+        ),
+        check(
+            assert_type(pd.read_excel(path, names=range(1), header=0), pd.DataFrame),
+            pd.DataFrame,
+        )
 
 
 def test_read_excel_io_types() -> None:

--- a/tests/test_io.py
+++ b/tests/test_io.py
@@ -909,6 +909,16 @@ def test_read_excel() -> None:
             ),
             pd.DataFrame,
         )
+        check(
+            assert_type(
+                pd.read_excel(
+                    path,
+                    usecols="A",
+                ),
+                pd.DataFrame,
+            ),
+            pd.DataFrame,
+        )
 
 
 def test_read_excel_io_types() -> None:

--- a/tests/test_io.py
+++ b/tests/test_io.py
@@ -52,6 +52,7 @@ import sqlalchemy.orm.decl_api
 from typing_extensions import assert_type
 
 from tests import (
+    TYPE_CHECKING_INVALID_USAGE,
     WINDOWS,
     check,
 )
@@ -285,6 +286,9 @@ def test_clipboard():
         ),
         DataFrame,
     )
+    if TYPE_CHECKING_INVALID_USAGE:
+        pd.read_clipboard(names="abcd")  # type: ignore[call-overload] # pyright: ignore[reportGeneralTypeIssues]
+        pd.read_clipboard(usecols="abcd")  # type: ignore[arg-type] # pyright: ignore[reportGeneralTypeIssues]
 
 
 def test_clipboard_iterator():
@@ -618,6 +622,10 @@ def test_types_read_csv() -> None:
             index_col=0,
         )
 
+        if TYPE_CHECKING_INVALID_USAGE:
+            pd.read_csv(path, names="abcd")  # type: ignore[call-overload] # pyright: ignore[reportGeneralTypeIssues]
+            pd.read_csv(path, usecols="abcd")  # type: ignore[arg-type] # pyright: ignore[reportGeneralTypeIssues]
+
         tfr1: TextFileReader = pd.read_csv(path, nrows=2, iterator=True, chunksize=3)
         tfr1.close()
 
@@ -733,6 +741,9 @@ def test_read_table():
             ),
             DataFrame,
         )
+        if TYPE_CHECKING_INVALID_USAGE:
+            pd.read_table(path, names="abcd")  # type: ignore[call-overload] # pyright: ignore[reportGeneralTypeIssues]
+            pd.read_table(path, usecols="abcd")  # type: ignore[arg-type] # pyright: ignore[reportGeneralTypeIssues]
 
 
 def test_read_table_iterator():
@@ -919,6 +930,8 @@ def test_read_excel() -> None:
             ),
             pd.DataFrame,
         )
+        if TYPE_CHECKING_INVALID_USAGE:
+            pd.read_excel(path, names="abcd")  # type: ignore[call-overload] # pyright: ignore[reportGeneralTypeIssues]
 
 
 def test_read_excel_io_types() -> None:

--- a/tests/test_io.py
+++ b/tests/test_io.py
@@ -219,6 +219,18 @@ def test_clipboard():
         DataFrame,
     )
     check(assert_type(read_clipboard(names=None), DataFrame), DataFrame)
+    check(
+        assert_type(read_clipboard(names=("first", "second"), header=0), DataFrame),
+        DataFrame,
+    )
+    check(assert_type(read_clipboard(names=range(2), header=0), DataFrame), DataFrame)
+    check(assert_type(read_clipboard(names=(1, "two"), header=0), DataFrame), DataFrame)
+    check(
+        assert_type(
+            read_clipboard(names=(("first", 1), ("last", 2)), header=0), DataFrame
+        ),
+        DataFrame,
+    )
 
 
 def test_clipboard_iterator():

--- a/tests/test_io.py
+++ b/tests/test_io.py
@@ -218,6 +218,7 @@ def test_clipboard():
         assert_type(read_clipboard(dtype=defaultdict(lambda: "f8")), DataFrame),
         DataFrame,
     )
+    check(assert_type(read_clipboard(names=None), DataFrame), DataFrame)
 
 
 def test_clipboard_iterator():

